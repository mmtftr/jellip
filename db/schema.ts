import { relations, sql } from "drizzle-orm";
import { text, integer, sqliteTable } from "drizzle-orm/sqlite-core";

export const levels = ["N5", "N4", "N3", "N2", "N1"] as const;
export const categories = [
  "grammar",
  "vocabulary",
  "kanji",
  "unknown",
] as const;
export const questions = sqliteTable("questions", {
  id: integer("id").primaryKey({ autoIncrement: true }),
  question: text("question").notNull(),
  category: text("category", {
    enum: categories,
  })
    .notNull()
    .default("unknown"),
  level: text("level", { enum: levels }).notNull().default("N5"),
  answers: text("answers", { mode: "json" }).$type<string[]>().notNull(),
  correctAnswer: integer("correctAnswer").notNull(),
  questionSet: integer("questionSet").references(() => questionSets.id),
});

export const questionRelations = relations(questions, ({ one, many }) => ({
  questionSet: one(questionSets, {
    fields: [questions.questionSet],
    references: [questionSets.id],
  }),
  userAnswers: many(answers),
}));

export const questionSets = sqliteTable("questionSets", {
  id: integer("id").primaryKey({ autoIncrement: true }),
  name: text("name").notNull(),
  description: text("description"),
});

export const questionSetRelations = relations(questionSets, ({ many }) => ({
  questions: many(questions),
}));

export const answers = sqliteTable("answers", {
  id: integer("id").primaryKey({ autoIncrement: true }),
  questionId: integer("questionId")
    .notNull()
    .references(() => questions.id),
  answer: integer("answer").notNull(),
  timestamp: text("date")
    .default(sql`(CURRENT_TIMESTAMP)`)
    .notNull(),
});

export const answersRelations = relations(answers, ({ one }) => ({
  question: one(questions, {
    fields: [answers.questionId],
    references: [questions.id],
  }),
}));
