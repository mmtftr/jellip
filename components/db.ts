import { drizzle } from "drizzle-orm/expo-sqlite";
import { openDatabaseSync } from "expo-sqlite/next";
import * as schema from "@/db/schema";

const expoDb = openDatabaseSync("db.db");
export const db = drizzle(expoDb, { schema });
