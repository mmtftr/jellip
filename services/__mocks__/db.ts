import { drizzle } from "drizzle-orm/better-sqlite3";
const Database = require("better-sqlite3");

const dir = require("os").tmpdir();
const dbName = dir + `/test-db${Math.floor(Math.random() * 10000)}.sqlite`;
const sqlite = new Database(dbName);

export const db = drizzle(sqlite);

console.log(dbName);
