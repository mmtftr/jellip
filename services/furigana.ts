import { useQuery, UseQueryResult } from "@tanstack/react-query";
import React, { useContext } from "react";

interface FuriganaArgs {
  sentence: string;
  enabled: boolean;
}
export type FuriganaPart = {
  text: string;
  furi?: string;
};

export interface FuriganaSentence {
  parts: FuriganaPart[];
}

type FuriganaHookReturn = UseQueryResult<FuriganaSentence, Error>;

export interface FuriganaService {
  getFurigana(sentence: string): Promise<FuriganaSentence>;
}

export const FuriganaContext = React.createContext({
  service: undefined as FuriganaService | undefined,
});

export const useFurigana = ({
  sentence,
  enabled,
}: FuriganaArgs): FuriganaHookReturn => {
  const { service } = useContext(FuriganaContext);

  const result = useQuery({
    queryKey: ["furigana", sentence, !!service],
    queryFn: () => {
      if (!service)
        return Promise.reject(Error("Furigana service not available"));

      return service.getFurigana(sentence);
    },
    enabled,
  });

  return result;
};
