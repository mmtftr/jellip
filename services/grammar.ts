// import { Question, QuestionWithAnswers } from "./questions";
// import Details from "../scripts/data/grammar_details.json";
// import Kuromoji from "@charlescoeder/react-native-kuromoji";
// import * as FileSystem from "expo-file-system";
// import { Asset } from "expo-asset";

// const kuromoji = Kuromoji.builder({
//   assets: {
//     "base.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/base.dat.gz")
//     ),
//     "cc.dat.gz": Asset.fromModule(require("../assets/kuromoji/dict/cc.dat.gz")),
//     "check.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/check.dat.gz")
//     ),
//     "tid_map.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/tid_map.dat.gz")
//     ),
//     "tid_pos.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/tid_pos.dat.gz")
//     ),
//     "tid.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/tid.dat.gz")
//     ),
//     "unk_char.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/unk_char.dat.gz")
//     ),
//     "unk_compat.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/unk_compat.dat.gz")
//     ),
//     "unk_invoke.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/unk_invoke.dat.gz")
//     ),
//     "unk_map.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/unk_map.dat.gz")
//     ),
//     "unk_pos.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/unk_pos.dat.gz")
//     ),
//     "unk.dat.gz": Asset.fromModule(
//       require("../assets/kuromoji/dict/unk.dat.gz")
//     ),
//   },
// }).build();

// const space = new RegExp("[（(]\\s+[)）]");
// const parenExpr = new RegExp("[（(].+[)）]");

// function processGrammar(text: string) {
//   return text
//     .split(",")
//     .map((s) => s.trim().replace("〜", "").replace(parenExpr, ""));
// }

// const DetailsSearchIndex: Record<string, number> = Details.reduce(
//   (acc, item) => ({
//     ...acc,
//     ...Object.fromEntries(
//       processGrammar(item.main_grammar).map((s) => [s, item.id])
//     ),
//   }),
//   {}
// );

// export const findRelatedGrammarPoints = async (question: Question) => {
//   const tok = await kuromoji;
//   let questionText = question.question;

//   if (space.test(questionText)) {
//     questionText = questionText.replace(
//       space,
//       question.answers[question.correctAnswer]
//     );
//   }

//   const grammars = new Set();

//   tok.tokenize(questionText).forEach((token: { surface_form: string }) => {
//     if (DetailsSearchIndex[token.surface_form]) {
//       grammars.add(DetailsSearchIndex[token.surface_form]);
//     }
//   });

//   return Array.from(grammars);
// };
