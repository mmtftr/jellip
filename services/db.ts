import { drizzle } from "drizzle-orm/expo-sqlite";
import { openDatabaseSync } from "expo-sqlite/next";
import * as schema from "@/db/schema";

const expoDb = openDatabaseSync("jlpt.db");
export const db = drizzle(expoDb, { schema });
