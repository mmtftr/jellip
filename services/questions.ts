import { db } from "@/services/db";
import { answers, questions } from "@/db/schema";
import { and, eq, inArray, sql } from "drizzle-orm";
import { openBrowserAsync } from "expo-web-browser";
import { canOpenURL, openURL } from "expo-linking";
import { answersTodayStore } from "./store";

export type Question = typeof questions.$inferSelect;
export type QuestionWithAnswers = typeof questions.$inferSelect & {
  userAnswers: (typeof answers.$inferSelect)[];
};
const getRandomQuestion = async ({
  categoryFilter,
  levelFilter,
}: {
  categoryFilter?: QuestionWithAnswers["category"][];
  levelFilter?: QuestionWithAnswers["level"][];
} = {}): Promise<QuestionWithAnswers | null> => {
  const question =
    (
      await db
        .select()
        .from(questions)
        .where(
          and(
            categoryFilter && inArray(questions.category, categoryFilter),
            levelFilter && inArray(questions.level, levelFilter)
          )
        )
        .orderBy(sql`random()`)
        .limit(1)
    )[0] || null;

  if (!question) return null;

  const answers = await getAnswers(question.id);

  return { ...question, userAnswers: answers };
};

const getAnswersToday = async () => {
  return db
    .select()
    .from(answers)
    .innerJoin(questions, eq(answers.questionId, questions.id))
    .where(eq(sql`date(${answers.timestamp})`, sql`date(current_timestamp)`));
};

const getAnswers = async (questionId: number) => {
  return db.select().from(answers).where(eq(answers.questionId, questionId));
};

const submitAnswer = async (questionId: number, answerIdx: number) => {
  answersTodayStore.getState().update((state) => state.val++);

  return db.insert(answers).values({ questionId, answer: answerIdx });
};

const lookupQuestion = async (question: Question) => {
  const lookupStr =
    (question?.question || "") + (question?.answers.join(" ") || "");

  if (__DEV__ || (await canOpenURL("shirabelookup://search")))
    return await openURL(
      `shirabelookup://search?w=${encodeURIComponent(lookupStr)}`
    );

  // fallback to Jisho
  openBrowserAsync(`https://jisho.org/search/${encodeURIComponent(lookupStr)}`);
};

const lookupAnswer = async (answer: string) => {
  openBrowserAsync(
    `https://www.google.com/search?q=${encodeURIComponent(
      (answer || "") + " 文法"
    )}`
  );
};

export {
  getRandomQuestion,
  getAnswersToday,
  getAnswers,
  submitAnswer,
  lookupAnswer,
  lookupQuestion,
};
