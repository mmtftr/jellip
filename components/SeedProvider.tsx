import { db } from "@/services/db"; // drizzle
import {
  answers,
  categories,
  levels,
  questionSets,
  questions,
} from "../db/schema";
import seedVals from "@/constants/seed.json";
import { useEffect, useState } from "react";
import { useAsyncStorage } from "@react-native-async-storage/async-storage";
import { ActivityIndicator } from "react-native";
import { getRandomQuestion } from "../services/questions";

const seed = async () => {
  try {
    const allValues = seedVals.questions.map(
      ({
        id,
        questionText: question,
        answers,
        correctAnswer,
        level,
        category,
        questionSet,
      }) => ({
        id,
        question,
        answers,
        category: categories.includes(category as any)
          ? (category as any)
          : "unknown",
        level: levels.includes(level as any) ? (level as any) : "N5",
        correctAnswer,
        questionSet,
      }),
    );
    console.log("Seeding database...");
    await db.insert(questionSets).values(seedVals.questionSets).execute();

    // insert 1K questions at a time.
    for (let i = 0; i < allValues.length / 1000; i++) {
      await db
        .insert(questions)
        .values(allValues.slice(i * 1000, (i + 1) * 1000))
        .execute();
    }
  } catch (e) {
    console.error(e);
  }
};

export const resetAndReseed = async () => {
  try {
    const answerObjects = await db.delete(answers).returning().execute();
    await db.delete(questions).execute();
    await db.delete(questionSets).execute();
    await seed();

    // try to restore answers
    await db.insert(answers).values(answerObjects).execute();
  } catch (e) {
    console.error(e);
  }
};

export default function SeedProvider({ children }: React.PropsWithChildren) {
  const [shouldSeed, setShouldSeed] = useState<boolean | null>(null);
  const seeded = useAsyncStorage("seeded");

  useEffect(() => {
    Promise.all([getRandomQuestion(), seeded.getItem()]).then(([q, val]) => {
      if (!q) return setShouldSeed(true);

      setShouldSeed(val === "true" ? false : true);
    });
  }, []);

  useEffect(() => {
    if (shouldSeed === true) {
      seed().then(() => {
        seeded.setItem("true");
        setShouldSeed(false);
      });
    }
  }, [shouldSeed]);
  if (shouldSeed !== false) {
    return <ActivityIndicator />;
  }

  return children;
}
