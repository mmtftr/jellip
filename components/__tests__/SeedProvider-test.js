import * as React from "react";
import {
  act,
  render,
  screen,
  waitFor,
  TestRenderer,
} from "@testing-library/react-native";
import * as z from "zod";
import {
  levels,
  categories,
  questions,
  questionSets,
  answers,
} from "../../db/schema";
import SeedProvider from "../SeedProvider";
import { db } from "@/services/db";
import { count, eq, sql } from "drizzle-orm";
import { Text } from "react-native";

jest.useFakeTimers();
jest.mock("@/services/db");
const { reset, migrate } = require("@/services/db");
jest.mock("@react-native-async-storage/async-storage");

describe("SeedProvider", () => {
  beforeAll(() => {
    migrate();
  });

  beforeEach(async () => {
    jest.clearAllMocks();
    await reset();
  });

  it("should render children after seed", async () => {
    render(
      <SeedProvider>
        <Text>Hi</Text>
      </SeedProvider>
    );

    expect(screen.queryByText("Hi")).toBeFalsy();

    await act(async () => {
      await jest.runAllTimersAsync();
    });

    expect(screen.getByText("Hi")).toBeTruthy();
  });

  it("should seed on empty db", async () => {
    expect((await db.select({ num: count() }).from(questions))[0].num).toBe(0);

    render(<SeedProvider />);

    await act(async () => {
      await jest.runAllTimers();
    });

    await waitFor(async () => {
      expect(
        (await db.select({ num: count() }).from(questions))[0].num
      ).toBeGreaterThan(0);
    });
  });

  it("should succeed even if duplicates exist", async () => {
    await db.insert(questionSets).values({ id: 1, name: "First" }).execute();
    await db
      .insert(questionSets)
      .values({ id: 2, name: "Set Without Questions" })
      .execute();
    await db
      .insert(questions)
      .values({
        question: "What is the capital of France?",
        answers: ["Paris", "London", "Berlin", "Madrid"],
        correctAnswer: 0,
        level: "easy",
        category: "geography",
        questionSet: 1,
      })
      .execute();
    await db
      .insert(questions)
      .values({
        question: "What is the capital of France?",
        answers: ["Paris", "London", "Berlin", "Madrid"],
        correctAnswer: 0,
        level: "easy",
        category: "geography",
        questionSet: 1,
      })
      .execute();

    render(<SeedProvider />);

    await act(async () => {
      await jest.runAllTimers();
    });

    await waitFor(async () => {
      expect(
        (
          await db
            .select({ num: count() })
            .from(questions)
            .where(eq(questions.questionSet, 2))
        )[0].num
      ).toBeGreaterThan(0);
    });
  });

  it("should seed missing question set", async () => {
    await db
      .insert(questionSets)
      .values({ id: 1, name: "Missing Set" })
      .execute();
    await db
      .insert(questions)
      .values({
        question: "What is the capital of France?",
        answers: ["Paris", "London", "Berlin", "Madrid"],
        correctAnswer: 0,
        level: "easy",
        category: "geography",
        questionSet: 1,
      })
      .execute();

    render(<SeedProvider />);

    await act(async () => {
      await jest.runAllTimers();
    });

    await waitFor(async () => {
      expect(
        (
          await db
            .select({ num: count() })
            .from(questions)
            .where(eq(questions.questionSet, 1))
        )[0].num
      ).toBe(1);

      expect(
        (await db.select({ num: count() }).from(questions))[0].num
      ).toBeGreaterThan(1);
    });
  });
});

describe("seed.json", () => {
  it("should be valid", () => {
    const seedFile = require("../../constants/seed.json");

    const questionSchema = z.array(
      z.object({
        id: z.number(),
        question: z.string(),
        answers: z.array(z.string()),
        correctAnswer: z.number(),
        level: z.union(levels.map(z.literal)),
        category: z.union(categories.map(z.literal)),
        questionSet: z.number(),
      })
    );
    questionSchema.parse(seedFile.questions);
  });
});
