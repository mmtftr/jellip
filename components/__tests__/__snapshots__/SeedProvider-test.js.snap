// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeedProvider should render correctly 1`] = `<ActivityIndicator />`;
